lib/cn/node_split.c

```diff
@@ -366,3 +370,85 @@ cn_tree_node_get_split_key(
 
     return err;
 }
+
+static void
+kvset_split_res_init(struct cn_compaction_work *w, struct kvset_split_res *result, uint ks_idx)
+{
+    memset(result, 0, sizeof(*result));
+
+    for (int i = LEFT; i <= RIGHT; i++) {
+        uint idx = (i == LEFT ? ks_idx : ks_idx + w->cw_kvset_cnt);
+
+        result->ks[i].blks = &w->cw_outv[idx];
+        blk_list_init(&result->ks[i].blks->kblks);
+        blk_list_init(&result->ks[i].blks->vblks);
+        result->ks[i].blks->bl_vused = 0;
+
+        result->ks[i].blks_commit = &w->cw_split.commit[idx];
+        blk_list_init(result->ks[i].blks_commit);
+
+        result->ks[i].vgmap = &w->cw_vgmap[idx];
+    }
+
+    result->blks_purge = &w->cw_split.purge[ks_idx];
+    blk_list_init(result->blks_purge);
+}
+
+static void
+kvset_split_res_free(struct kvset *ks, struct kvset_split_res *result)
+{
+    for (int i = LEFT; i <= RIGHT; i++) {
+        blk_list_free(&result->ks[i].blks->kblks);
+        blk_list_free(&result->ks[i].blks->vblks);
+        blk_list_free(result->ks[i].blks_commit);
+    }
+
+    blk_list_free(result->blks_purge);
+}
+
+merr_t
+cn_split(struct cn_compaction_work *w)
+{
+    struct kvset_list_entry *le;
+    struct key_obj split_kobj;
+    merr_t err;
+    uint i;
+
+    err = cn_tree_node_get_split_key(w->cw_node, w->cw_split.key, HSE_KVS_KEY_LEN_MAX,
+                                     &w->cw_split.klen);
+    if (err)
+        return err;
+
+    key2kobj(&split_kobj, w->cw_split.key, w->cw_split.klen);
+
+    for (i = 0, le = list_first_entry_or_null(&w->cw_node->tn_kvset_list, typeof(*le), le_link);
+         i < w->cw_kvset_cnt;
+         i++, le = list_next_entry(le, le_link)) {
```

Comment:
Fully agreed. I was actually thinking about it this morning and wanted to protect against the scheduler calling node split with a subset of kvsets in a node.

Reply:
Better make it a live check, in the past we could add kvsets to nodes while they were being compacted...

Reply:
yes, I'm adding a live check.

Reply:
return EBUG if not assert?

Reply:
I'm thinking to add the below check in cn_tree_prepare_compaction() and fail much earlier with EBUG. Guess that should work fine?

```
if (cn_ns_kvsets(&tn->tn_ns) != w->cw_kvset_cnt)
    return merr(EBUG);
```

Reply:
Did you put something else in place to prevent a spill while you're doing a split?

Reply:
Nope. Alex mentioned that this synchronization can be achieved using the per-node compaction token and will be taken care of in your upcoming tree maintenance PR.

Reply:
As you're already aware of, the compaction token does prevent a k/kv-compact to be scheduled on a node that's split.

We need to change the spill logic to grab the compaction token on a child node before starting to spill to this node. This token must be released after the spill to this node has been committed to cNDB and is updated with the new kvset. Guess that should work?

The description above assumes Gaurav's upcoming changes to spill are in place.

Reply:
Something along those lines.  But the current token logic seems a bit broken?  There's code to say "use_token", that then results in a CAS to get the token, but doesn't fail if the CAS fails.  Maybe just using a RW lock with the following rules:
- split and join need write lock
- spill to leaf, k-compact and kv-compact need read lock

scheduler would of course have to keep k-compacts and kv-compacts from working on the same kvsets, but it already does that with `kvset_get_workid(struct kvset *ks)`.

Reply:
re what i said was broken, in cn_comp_work_init():
```
    w->cw_have_token = use_token;
    if (w->cw_have_token)
        cn_node_comp_token_get(tn);
```
i didn't look wider, but by itself that seems wrong.   `use_token` seems like a request to use the token, but result of the CAS to get the token isn't set.

Reply:
cn_comp_work_init() is a test code in cn_tree_test.c file. However, it does look funny that it attempts to get a token if it already has it although no harm.

Reply:
Heh.  I cscope my way around w/o really paying attention to the files I land in.  :)

Reply:
I think use of the token is correct insofar as what it does (i.e., ensure exclusive access to the node to the token holder).
What it never did was prevent new kvsets being prepended to the node from a parent spill, as Nabeel pointed out.
I need to think about this...

Reply:
Correct.  That was not necessary in the past because nodes were never removed from the tree.  Current thinking is this will be handled by the root spill loop:  when it iterates to the next leaf node, it will try to get token, if it can't get the token, it has to wait and try again (w/o holding a ref to the node bc it may be split or joined while waiting for token).